File: Kogebog.API/Kogebog.API.Repository/Database/Entities/RecipeIngredient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Kogebog.API.Repository.Database.Entities
{
    public class RecipeIngredient
    {
        public Guid Id { get; set; }
        public double Quantity { get; set; }
        public Guid UnitId { get; set; }
        public Unit? Unit { get; set; }
        public Guid RecipeId { get; set; }
        public Recipe? Recipe { get; set; }
        public Guid IngredientId { get; set; }
        public Ingredient? Ingredient { get; set; }
    }
}

File: Kogebog.App/kogebog_flutter/lib/models/recipe_ingredient.dart
import 'package:kogebog_flutter/models/ingredient.dart';
import 'package:kogebog_flutter/models/recipe.dart';
import 'package:kogebog_flutter/models/unit.dart';

class RecipeIngredient {
  final String id;
  final double quantity;
  final String? unitId;
  final Unit? unit;
  final String? recipeId;
  final Recipe? recipe;
  final String? ingredientId;
  final Ingredient? ingredient;

  RecipeIngredient({required this.id, required this.quantity, this.unitId, this.unit, this.recipeId, this.recipe, this.ingredientId, this.ingredient});

  factory RecipeIngredient.fromJson(Map<String, dynamic> json) {
  try {
    return RecipeIngredient(
      id: json['id'] ?? '',
      quantity: json['quantity'] != null 
          ? (json['quantity'] is int 
              ? (json['quantity'] as int).toDouble() 
              : json['quantity'])
          : 0.0,
      unit: json['unit'] != null ? Unit.fromJson(json['unit']) : null,
      ingredient: json['ingredient'] != null ? Ingredient.fromJson(json['ingredient']) : null,
      // Don't try to parse recipe here to avoid circularity
      recipeId: json['recipeId'],
    );
  } catch (e) {
    print("Error in RecipeIngredient.fromJson: $e");
    return RecipeIngredient(
      id: json['id'] ?? 'error',
      quantity: 0.0,
    );
  }
}

  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "quantity": "${quantity == quantity.toInt() ? quantity.toInt() : quantity.toStringAsFixed(2)}",
      "unitId": unitId,
      "unit": unit,
      "recipeId": recipeId,
      "recipe": recipe,
      "ingredientId": ingredientId,
      "ingredient": ingredient
    };
  }
}